experimental/vindex/map.go

```diff
@@ -86,68 +93,172 @@ func (b IndexBuilder) pullFromDatabase(ctx context.Context, start uint64) {
 				klog.Exitf("Panic: failed to read leaf at index %d: %v", i, err)
 			}
 			hashes := b.mapFn(l.Leaf)
-			if err := b.addIndex(i, hashes); err != nil {
+			if err := b.wal.append(i, hashes); err != nil {
 				klog.Exitf("failed to add index to entry for leaf %d: %v", i, err)
 			}
+			// TODO(mhutchinson): announce updates to map construction code
 		}
 	}
 
 	// TODO(mhutchinson): the raw log checkpoint needs to be propagated into the map checkpoint
 	_ = rawCp
 }
 
-func (b IndexBuilder) addIndex(idx uint64, hashes [][]byte) error {
-	// TODO(mhutchinson): this should also push this off to the map construction code
-	return b.wal.append(idx, hashes)
-}
-
 type writeAheadLog struct {
 	walPath string
-
-	entries []string
+	f       *os.File
 }
 
-// init reads the file and determines what the last mapped log index was, and returns it.
-// This method populates entries with the lines from the WAL up to and including the last
-// good entry.
+// init verifies that the log is in good shape, and returns the index that is expected next.
+// It also opens the log for appending to.
+//
+// Note that it returns the next expected index to avoid awkwardness with the meaning of 0,
+// which could mean 0 was successfully read from a previous run, or that there was no log.
 func (l *writeAheadLog) init() (uint64, error) {
+	idx, err := l.validate()
+
+	ffs := os.O_WRONLY | os.O_APPEND
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return idx, err
+		}
+		ffs |= os.O_CREATE
```

Comment:
`O_CREATE` will only attempt to create a file if it doesn't exist, were you thinking of `O_EXCL`?

As it stands, I think two instances of this binary running concurrently would both successfully open the file and start appending to it (not _super_ clear what the resuling contents of the file would be in that case - I think it might be kernel/filesystem dependent)